<a href="https://colab.research.google.com/github/Minjoong1999/Sentimental-Analysis/blob/main/%EC%A0%84%EC%A2%85%EC%84%A4_%EA%B0%90%EC%84%B1%EB%B6%84%EC%84%9D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Fine-tune the KcBERT model on the "/content/Train_set.csv" dataset, splitting it into stratified training and testing sets, and evaluate the model's performance on the test set using metrics such as Accuracy and F1-score.

## Install necessary libraries

### Subtask:
Install transformers, datasets, and other required libraries.


**Reasoning**:
Install the required libraries using pip.



In [ ]:
!pip install transformers datasets pandas scikit-learn

## Load the dataset

### Subtask:
Load the `Train_set.csv` file into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the `Train_set.csv` file into a pandas DataFrame and display the head of the dataframe to confirm successful loading.



In [ ]:
import pandas as pd

df = pd.read_csv('/content/Train_set.csv')
display(df.head())

파일명  \
0          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
1          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
2     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
3     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
4  [시승기] 제네시스 첫 전용 전기차 GV60 '후두부를 강타하는 기막힌 반전'.docx   

                                                  문장  label  
0  필기 인식의 손글씨 입력을 통해 사용되지만 대시보드 상단 대형 디스플레이를 통해 대...      3  
1  다만 입체적 패턴과 화려함을 전달했던 디지털 계기판과 다양한 정보와 시인성 높은 헤...      5  
2  계기판은 앞서 티볼리와 코란도를 통해 봐왔던 3분할 디지털 클러스터가 적용되고 대신...      4  
3  센터페시아 상단에는 12.3인치 터치식 디지털 디스플레이가 자리하고 또 그 아래에는...      3  
4  이 밖에 실내에선 아이오닉 5를 통해서도 봤던 디지털 계기판과 센터 디스플레이를 파...      5

## Preprocess the data

### Subtask:
Clean and prepare the text data for model training (e.g., handle missing values, tokenize).


**Reasoning**:
Check for missing values in the '문장' column and remove leading/trailing whitespace from the text data.



In [ ]:
print("Missing values before cleaning:")
print(df['문장'].isnull().sum())

df['문장'] = df['문장'].str.strip()

print("\nMissing values after cleaning:")
print(df['문장'].isnull().sum())
display(df.head())

Missing values before cleaning:
0

Missing values after cleaning:
0


파일명  \
0          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
1          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
2     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
3     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
4  [시승기] 제네시스 첫 전용 전기차 GV60 '후두부를 강타하는 기막힌 반전'.docx   

                                                  문장  label  
0  필기 인식의 손글씨 입력을 통해 사용되지만 대시보드 상단 대형 디스플레이를 통해 대...      3  
1  다만 입체적 패턴과 화려함을 전달했던 디지털 계기판과 다양한 정보와 시인성 높은 헤...      5  
2  계기판은 앞서 티볼리와 코란도를 통해 봐왔던 3분할 디지털 클러스터가 적용되고 대신...      4  
3  센터페시아 상단에는 12.3인치 터치식 디지털 디스플레이가 자리하고 또 그 아래에는...      3  
4  이 밖에 실내에선 아이오닉 5를 통해서도 봤던 디지털 계기판과 센터 디스플레이를 파...      5

## Split the dataset

### Subtask:
Split the data into stratified training and testing sets.


**Reasoning**:
Split the DataFrame into stratified training and testing sets using train_test_split, ensuring the split is stratified based on the 'label' column.



In [ ]:
from sklearn.model_selection import train_test_split

train_df, test_df = train_test_split(df, test_size=0.2, random_state=42, stratify=df['label'])

print("Training set shape:", train_df.shape)
print("Testing set shape:", test_df.shape)
display(train_df.head())
display(test_df.head())

Training set shape: (424, 3)
Testing set shape: (106, 3)


파일명  \
294  모든 것을 담다, 메르세데스-벤츠 6세대 C300 AMG 라인 시승기.docx   
90              제네시스 2세대 G90 3.5 터보 AWD 시승기.docx   
77        새로운 길을 간다. 현대 7세대 아반떼 1.6 IVT 시승기.docx   
308                               XM3_bobae.docx   
526                        제네시스 G70 리뷰(클리앙).docx   

                                                    문장  label  
294  대시보드는 12.3인치의 LCD 계기판과 운전자 쪽으로 6도 기울어진 11.9인치 ...      3  
90             G80과 GV80의 그것과는 달리 디스플레이창의 나란히 배치되어 있다.      3  
77   K9은 LG디스플레이가 만들어 LG전자가 납품한 것이지만 제네시스 GV80의 3D디...      3  
308  계기판은 10.25인치 tft클러스터가 들어가고\n9.3인치 이지커넥터 네비게이션\...      3  
526   헤드업디스플레이(HUD)는 옵션으로 추가 가능한데 BMW 것보다는 역시 좀 떨어집니다.      2

파일명  \
33                     [시승기] 기아 2021 K3.docx   
515                     기아 스포티지 리뷰(클리앙).docx   
373             Carnival_hybrid_Youtube.docx   
235  쌍용차 리스펙 티볼리, 인포콘 하나로 체질을 바꾼 소형 SUV.docx   
122             모든 것이 변했다, 쌍용 뉴 코란도 시승기.docx   

                                                    문장  label  
33   대신 가운데 디스플레이 테두리 안쪽 아래에 각종 기능 버튼이 터치식으로 자리를 잡으...      5  
515         특히 전 바늘계기판을 좋아하다 보니 소나타의 실내의 만족도 무척 높았습니다.      5  
373                  그리고 앞차와의 거리도 요즘 헤더 디스플레이 통해서 보여줘요      4  
235  한편 리스펙 티볼리는 또 스마트 폰 미러링, 10.25인치 풀 디지털 클러스터, 9...      4  
122                  가장 먼저 눈에 띄는 부분은 10.25인치의 디지털 계기판.      3

## Load the pre-trained model

### Subtask:
Load the KcBERT model and tokenizer using the transformers library.


**Reasoning**:
Import the necessary classes and load the tokenizer and model for KcBERT.



In [ ]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification

model_name = "beomi/kcbert-base"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=6) # Assuming 6 labels based on previous steps

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/250k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at beomi/kcbert-base and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


## Prepare data for training

### Subtask:
Tokenize and format the data for input into the model.


**Reasoning**:
Define a function to tokenize the text and encode the labels, then apply it to the training and testing DataFrames and convert them to Hugging Face Dataset objects.



In [ ]:
from datasets import Dataset

def tokenize_function(row):
    inputs = tokenizer(row['문장'], padding='max_length', truncation=True)
    inputs['labels'] = row['label']
    return inputs

train_dataset = Dataset.from_pandas(train_df)
test_dataset = Dataset.from_pandas(test_df)

tokenized_train_dataset = train_dataset.map(tokenize_function, batched=True)
tokenized_test_dataset = test_dataset.map(tokenize_function, batched=True)

print("Tokenized training dataset:")
print(tokenized_train_dataset)
print("\nTokenized testing dataset:")
print(tokenized_test_dataset)

Map:   0%|          | 0/424 [00:00<?, ? examples/s]

Map:   0%|          | 0/106 [00:00<?, ? examples/s]

Tokenized training dataset:
Dataset({
    features: ['파일명', '문장', 'label', '__index_level_0__', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
    num_rows: 424
})

Tokenized testing dataset:
Dataset({
    features: ['파일명', '문장', 'label', '__index_level_0__', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
    num_rows: 106
})


## Fine-tune the model

### Subtask:
Train the KcBERT model on the training set.


**Reasoning**:
Train the KcBERT model on the training set using the Hugging Face Trainer.



**Reasoning**:
The previous command failed because `evaluation_strategy` is not a valid argument for `TrainingArguments` in the current version of the transformers library. The argument should be `eval_strategy`. I will correct the argument name and rerun the training code.



In [ ]:
from transformers import TrainingArguments, Trainer

training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=3,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    warmup_steps=500,
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
    eval_strategy="epoch",  # Corrected argument name
    report_to="none", # Disable W&B integration
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train_dataset,
    eval_dataset=tokenized_test_dataset,
)

trainer.train()

TrainOutput(global_step=159, training_loss=1.4510461939205914, metrics={'train_runtime': 5010.7601, 'train_samples_per_second': 0.254, 'train_steps_per_second': 0.032, 'total_flos': 196107001257600.0, 'train_loss': 1.4510461939205914, 'epoch': 3.0})

**Reasoning**:
The traceback indicates a `ValueError` related to the W&B API key length. This is because the `Trainer` by default tries to integrate with Weights & Biases, which requires authentication. Since W&B is not necessary for this specific subtask (training the model), I will disable the W&B integration in the `TrainingArguments`.



# Task
Preprocess the vehicle cluster reviews stored in "/content/reviews.csv" by removing duplicate sentences, special characters, and sentences that are too long. Save the preprocessed data to an Excel file.

## Load the data

### Subtask:
Load the data from "/content/reviews.csv" into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the data from the CSV file into a DataFrame. Then, display the head and info of the DataFrame.



In [ ]:
import pandas as pd

reviews_df = pd.read_csv('/content/reviews.csv')
display(reviews_df.head())
reviews_df.info()

파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...  
1                       안정성은 기본.계기판 디자인만 빼고 완전 만족!!!  
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...  
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다(...  
4            디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 ㅂㄹ안좋음

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1879 entries, 0 to 1878
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   파일명     1879 non-null   object
 1   문장      1879 non-null   object
dtypes: object(2)
memory usage: 29.5+ KB


## Handle duplicates

### Subtask:
Remove duplicate sentences from the DataFrame.


**Reasoning**:
Remove duplicate rows based on the '문장' column and print the shape before and after removing duplicates.



In [ ]:
print("Shape before removing duplicates:", reviews_df.shape)
reviews_df.drop_duplicates(subset=['문장'], inplace=True)
print("Shape after removing duplicates:", reviews_df.shape)
display(reviews_df.head())

Shape before removing duplicates: (1879, 2)
Shape after removing duplicates: (1769, 2)


파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...  
1                       안정성은 기본.계기판 디자인만 빼고 완전 만족!!!  
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...  
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다(...  
4            디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 ㅂㄹ안좋음

## Remove special characters

### Subtask:
Clean the text data by removing special characters.


**Reasoning**:
Define a function to remove special characters and apply it to the '문장' column.



In [ ]:
import re

def remove_special_characters(text):
    text = re.sub(r'[^가-힣a-zA-Z0-9\s]', '', text)
    return text

reviews_df['문장'] = reviews_df['문장'].apply(remove_special_characters)
display(reviews_df.head())

파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...  
1                           안정성은 기본계기판 디자인만 빼고 완전 만족  
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...  
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다약...  
4              디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 안좋음

## Filter long sentences

### Subtask:
Remove sentences that exceed a certain length threshold.


**Reasoning**:
Define a function to filter sentences by length, apply it to the DataFrame, filter the DataFrame, print shapes, and display the head of the filtered DataFrame.



In [ ]:
def filter_sentence_length(sentence, max_length=200):
    return len(sentence) <= max_length

print("Shape before filtering by length:", reviews_df.shape)

length_mask = reviews_df['문장'].apply(filter_sentence_length)
reviews_df = reviews_df[length_mask]

print("Shape after filtering by length:", reviews_df.shape)
display(reviews_df.head())

Shape before filtering by length: (1769, 2)
Shape after filtering by length: (1755, 2)


파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...  
1                           안정성은 기본계기판 디자인만 빼고 완전 만족  
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...  
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다약...  
4              디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 안좋음

## Save the processed data

### Subtask:
Save the preprocessed data to a new Excel file.


**Reasoning**:
Save the preprocessed DataFrame to an Excel file without the index.



In [ ]:
reviews_df.to_excel('preprocessed_reviews.xlsx', index=False)

## Summary:

### Data Analysis Key Findings

*   The initial dataset contained 1879 reviews with two columns: '파일명' and '문장'.
*   Removing duplicate sentences reduced the number of reviews from 1879 to 1769.
*   Special characters were successfully removed from the '문장' column.
*   Filtering out sentences longer than 200 characters further reduced the number of reviews from 1769 to 1755.
*   The preprocessed data, containing 1755 cleaned and filtered reviews, was saved to an Excel file.

### Insights or Next Steps

*   The preprocessing steps successfully cleaned and filtered the review data, preparing it for further analysis, such as sentiment analysis or topic modeling.
*   The threshold for sentence length (currently 200 characters) could be adjusted based on the specific requirements of subsequent analysis steps.


# Task
Perform sentiment analysis on the '문장' column of the "/content/preprocessed_reviews.xlsx" file using the "/content/Drive/kcbert_selftuned모델" model, save the results to an Excel file, and extract and return sentences with similar softmax probabilities.

## Load the preprocessed data

### Subtask:
Load the `preprocessed_reviews.xlsx` file into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the preprocessed reviews data from the Excel file into a DataFrame. Then display the head and info of the DataFrame.



In [ ]:
import pandas as pd

preprocessed_reviews_df = pd.read_excel('/content/preprocessed_reviews.xlsx')
display(preprocessed_reviews_df.head())
preprocessed_reviews_df.info()

파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...  
1                           안정성은 기본계기판 디자인만 빼고 완전 만족  
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...  
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다약...  
4              디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 안좋음

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1754 entries, 0 to 1753
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   파일명     1754 non-null   object
 1   문장      1754 non-null   object
dtypes: object(2)
memory usage: 27.5+ KB


## Load the fine-tuned model and tokenizer

### Subtask:
Load the fine-tuned KcBERT model and tokenizer from the specified local path.


**Reasoning**:
Import the necessary classes and load the tokenizer and model for KcBERT from the specified local path.



In [ ]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from google.colab import drive
import os

# Mount Google Drive if not already mounted
if not os.path.exists('/content/Drive'):
    drive.mount('/content/Drive')
else:
    print("Google Drive already mounted.")


model_path = "/content/fine_tuned_kcbert"
tokenizer = AutoTokenizer.from_pretrained(model_path)
model = AutoModelForSequenceClassification.from_pretrained(model_path)

Google Drive already mounted.


In [ ]:
def tokenize_sentence(sentence):
    inputs = tokenizer(sentence, padding=True, truncation=True, return_tensors="pt")
    return inputs

tokenized_reviews = preprocessed_reviews_df['문장'].apply(tokenize_sentence)

In [ ]:
import torch

def predict_sentiment(tokenized_input):
    with torch.no_grad():
        outputs = model(**tokenized_input)
    logits = outputs.logits
    probabilities = torch.softmax(logits, dim=1).squeeze().tolist()
    return probabilities

sentiment_probabilities = tokenized_reviews.apply(predict_sentiment)

In [ ]:
def is_probability_similar(probabilities, threshold=0.1):
    sorted_probs = sorted(probabilities, reverse=True)
    return (sorted_probs[0] - sorted_probs[1]) < threshold

similar_probabilities_mask = sentiment_probabilities.apply(is_probability_similar)

print("Number of sentences with similar probabilities:", similar_probabilities_mask.sum())

Number of sentences with similar probabilities: 1752


In [ ]:
similar_probability_sentences_df = preprocessed_reviews_df[similar_probabilities_mask].copy()
similar_probability_sentences_df['softmax_probabilities'] = sentiment_probabilities[similar_probabilities_mask]

similar_probability_sentences_df.to_excel('similar_probability_sentences.xlsx', index=False)

print("\nSentences with similar probabilities extracted and saved to 'similar_probability_sentences.xlsx'")
display(similar_probability_sentences_df.head())


Sentences with similar probabilities extracted and saved to 'similar_probability_sentences.xlsx'


파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  \
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...   
1                           안정성은 기본계기판 디자인만 빼고 완전 만족   
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...   
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다약...   
4              디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 안좋음   

                               softmax_probabilities  
0  [0.18204650282859802, 0.18858492374420166, 0.1...  
1  [0.23855803906917572, 0.1573256254196167, 0.17...  
2  [0.17449986934661865, 0.18549244105815887, 0.1...  
3  [0.1807439923286438, 0.2068440169095993, 0.098...  
4  [0.16904684901237488, 0.18811923265457153, 0.1...

In [ ]:
preprocessed_reviews_df['softmax_probabilities'] = sentiment_probabilities
# Assuming you want to add predicted labels as well, though not explicitly in the plan step
# You would need to get the predicted labels from the softmax_probabilities if needed
# For now, just saving probabilities as requested by the overall task

preprocessed_reviews_df.to_excel('reviews_with_probabilities.xlsx', index=False)

print("\nOriginal reviews with softmax probabilities saved to 'reviews_with_probabilities.xlsx'")


Original reviews with softmax probabilities saved to 'reviews_with_probabilities.xlsx'


# Task
Analyze the '문장' column in "/content/preprocessed_reviews.xlsx" using the "/content/Drive/kcbert_selftuned" and "/content/fine_tuned_kcbert" models to perform sentiment analysis. Save the results to an Excel file. Additionally, extract sentences with similar softmax probabilities into a separate Excel file and return them. Also, extract sentences with confident (not similar) softmax probabilities into another separate Excel file and return them. Accept all steps without asking for confirmation.

## Load the data with probabilities

### Subtask:
Load the `reviews_with_probabilities.xlsx` file into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the data from the Excel file into a DataFrame, then display the head and info of the DataFrame.



In [ ]:
reviews_with_probabilities_df = pd.read_excel('reviews_with_probabilities.xlsx')
display(reviews_with_probabilities_df.head())
reviews_with_probabilities_df.info()

파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  \
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...   
1                           안정성은 기본계기판 디자인만 빼고 완전 만족   
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...   
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다약...   
4              디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 안좋음   

                               softmax_probabilities  
0  [0.18204650282859802, 0.18858492374420166, 0.1...  
1  [0.23855803906917572, 0.1573256254196167, 0.17...  
2  [0.17449986934661865, 0.18549244105815887, 0.1...  
3  [0.1807439923286438, 0.2068440169095993, 0.098...  
4  [0.16904684901237488, 0.18811923265457153, 0.1...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1754 entries, 0 to 1753
Data columns (total 3 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   파일명                    1754 non-null   object
 1   문장                     1754 non-null   object
 2   softmax_probabilities  1754 non-null   object
dtypes: object(3)
memory usage: 41.2+ KB


## Extract sentences with confident probabilities

### Subtask:
Extract sentences where the softmax probabilities are not similar (confident predictions).

**Reasoning**:
Use the inverse of the `similar_probabilities_mask` to filter the DataFrame and extract sentences with confident softmax probabilities. Save the filtered DataFrame to a new Excel file and display the head.

In [ ]:
confident_probabilities_mask = ~similar_probabilities_mask

print("Number of sentences with confident probabilities:", confident_probabilities_mask.sum())

confident_probability_sentences_df = preprocessed_reviews_df[confident_probabilities_mask].copy()
confident_probability_sentences_df['softmax_probabilities'] = sentiment_probabilities[confident_probabilities_mask]

confident_probability_sentences_df.to_excel('confident_probability_sentences.xlsx', index=False)

print("\nSentences with confident probabilities extracted and saved to 'confident_probability_sentences.xlsx'")
display(confident_probability_sentences_df.head())

Number of sentences with confident probabilities: 2

Sentences with confident probabilities extracted and saved to 'confident_probability_sentences.xlsx'


파일명  \
758               K5_Youtube.docx   
951  casper_electric_Youtube.docx   

                                                    문장  \
758  그러니까 현대는 약간 기존 느낌에 더 가까운 것 같은 동그란 계기판을 사용하고 있고...   
951                            네 그리고 클러스터를 잠깐 한번 살펴볼게요   

                                 softmax_probabilities  
758  [0.17146332561969757, 0.16635064780712128, 0.1...  
951  [0.16167305409908295, 0.14812591671943665, 0.1...

## Filter short sentences

### Subtask:
Remove sentences that are too short.

**Reasoning**:
Define a function to filter sentences by minimum length, apply it to the DataFrame, filter the DataFrame, print shapes, and display the head of the filtered DataFrame.

In [ ]:
def filter_short_sentences(sentence, min_length=10):
    return len(sentence) >= min_length

print("Shape before filtering by length:", preprocessed_reviews_df.shape)

short_length_mask = preprocessed_reviews_df['문장'].apply(filter_short_sentences)
filtered_reviews_df = preprocessed_reviews_df[short_length_mask]

print("Shape after filtering by length:", filtered_reviews_df.shape)
display(filtered_reviews_df.head())

Shape before filtering by length: (1754, 3)
Shape after filtering by length: (1752, 3)


파일명  \
0  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
1  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
2  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
3  2025 쉐보레 트랙스 크로스오버 리뷰.docx   
4  2025 쉐보레 트랙스 크로스오버 리뷰.docx   

                                                  문장  \
0  옵션 또한 부족하다고 느껴지지않고 다만 디스플레이의 활용성이 낮아 계기판의 정보가 ...   
1                           안정성은 기본계기판 디자인만 빼고 완전 만족   
2  그 계기판이 작고 rpm보이는 메뉴가 한 가지 뿐이라 불편함 연비는 안 좋은 거 같...   
3  그리고 25년형을 구매했음에도 계기판과 디스플레이가 안들어오는 현상을 경험했습니다약...   
4              디스플레이 반응속도 느림 내장재 별로 삐걱삐걱 소리남 연비도 안좋음   

                               softmax_probabilities  
0  [0.18204650282859802, 0.18858492374420166, 0.1...  
1  [0.23855803906917572, 0.1573256254196167, 0.17...  
2  [0.17449986934661865, 0.18549244105815887, 0.1...  
3  [0.1807439923286438, 0.2068440169095993, 0.098...  
4  [0.16904684901237488, 0.18811923265457153, 0.1...

## Save the filtered data

### Subtask:
Save the filtered data to a new Excel file.

**Reasoning**:
Save the filtered DataFrame to an Excel file without the index.

In [ ]:
filtered_reviews_df.to_excel('filtered_preprocessed_reviews.xlsx', index=False)
print("\nFiltered reviews saved to 'filtered_preprocessed_reviews.xlsx'")


Filtered reviews saved to 'filtered_preprocessed_reviews.xlsx'


In [ ]:
removed_sentences_count = preprocessed_reviews_df.shape[0] - filtered_reviews_df.shape[0]
print(f"제거된 문장 수: {removed_sentences_count}")

제거된 문장 수: 2


# Task
Fine-tune the KcBERT base model using the data in "/content/Train_set.csv", save the fine-tuned model as "/content/kcbert_selftuned", then load the data from "/content/preprocessed_reviews.xlsx", remove sentences shorter than 10 characters from the '문장' column, perform sentiment analysis on the remaining sentences using the fine-tuned model, identify sentences with similar softmax probabilities, report the count of such sentences, and save them to an Excel file.

## Load the training data

### Subtask:
Load the data from "/content/Train_set.csv" into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the `Train_set.csv` file into a pandas DataFrame and display the head of the dataframe to confirm successful loading.



In [ ]:
import pandas as pd

df = pd.read_csv('/content/Train_set.csv')
display(df.head())

파일명  \
0          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
1          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
2     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
3     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
4  [시승기] 제네시스 첫 전용 전기차 GV60 '후두부를 강타하는 기막힌 반전'.docx   

                                                  문장  label  
0  필기 인식의 손글씨 입력을 통해 사용되지만 대시보드 상단 대형 디스플레이를 통해 대...      3  
1  다만 입체적 패턴과 화려함을 전달했던 디지털 계기판과 다양한 정보와 시인성 높은 헤...      5  
2  계기판은 앞서 티볼리와 코란도를 통해 봐왔던 3분할 디지털 클러스터가 적용되고 대신...      4  
3  센터페시아 상단에는 12.3인치 터치식 디지털 디스플레이가 자리하고 또 그 아래에는...      3  
4  이 밖에 실내에선 아이오닉 5를 통해서도 봤던 디지털 계기판과 센터 디스플레이를 파...      5

## Preprocess the training data

### Subtask:
Clean the text data for model training (e.g., handle missing values, tokenize).


**Reasoning**:
Check for missing values, strip whitespace, and display the head of the dataframe to show the results.



In [ ]:
print("Missing values before cleaning:")
print(df['문장'].isnull().sum())

df['문장'] = df['문장'].str.strip()

print("\nMissing values after cleaning:")
print(df['문장'].isnull().sum())
display(df.head())

Missing values before cleaning:
0

Missing values after cleaning:
0


파일명  \
0          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
1          제네시스 GV80, 절반의 합격점. 좋은데 부담스러운 첨단 기능.docx   
2     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
3     [시승기] '보급형 랜드로버 느낌' 부활 신호에서 확신으로 쌍용차 토레스.docx   
4  [시승기] 제네시스 첫 전용 전기차 GV60 '후두부를 강타하는 기막힌 반전'.docx   

                                                  문장  label  
0  필기 인식의 손글씨 입력을 통해 사용되지만 대시보드 상단 대형 디스플레이를 통해 대...      3  
1  다만 입체적 패턴과 화려함을 전달했던 디지털 계기판과 다양한 정보와 시인성 높은 헤...      5  
2  계기판은 앞서 티볼리와 코란도를 통해 봐왔던 3분할 디지털 클러스터가 적용되고 대신...      4  
3  센터페시아 상단에는 12.3인치 터치식 디지털 디스플레이가 자리하고 또 그 아래에는...      3  
4  이 밖에 실내에선 아이오닉 5를 통해서도 봤던 디지털 계기판과 센터 디스플레이를 파...      5

## Split the training data

### Subtask:
Split the data into stratified training and testing sets.


**Reasoning**:
Split the DataFrame into stratified training and testing sets using train_test_split, ensuring the split is stratified based on the 'label' column.



In [ ]:
from sklearn.model_selection import train_test_split

train_df, test_df = train_test_split(df, test_size=0.2, random_state=42, stratify=df['label'])

print("Training set shape:", train_df.shape)
print("Testing set shape:", test_df.shape)
display(train_df.head())
display(test_df.head())

Training set shape: (424, 3)
Testing set shape: (106, 3)


파일명  \
294  모든 것을 담다, 메르세데스-벤츠 6세대 C300 AMG 라인 시승기.docx   
90              제네시스 2세대 G90 3.5 터보 AWD 시승기.docx   
77        새로운 길을 간다. 현대 7세대 아반떼 1.6 IVT 시승기.docx   
308                               XM3_bobae.docx   
526                        제네시스 G70 리뷰(클리앙).docx   

                                                    문장  label  
294  대시보드는 12.3인치의 LCD 계기판과 운전자 쪽으로 6도 기울어진 11.9인치 ...      3  
90             G80과 GV80의 그것과는 달리 디스플레이창의 나란히 배치되어 있다.      3  
77   K9은 LG디스플레이가 만들어 LG전자가 납품한 것이지만 제네시스 GV80의 3D디...      3  
308  계기판은 10.25인치 tft클러스터가 들어가고\n9.3인치 이지커넥터 네비게이션\...      3  
526   헤드업디스플레이(HUD)는 옵션으로 추가 가능한데 BMW 것보다는 역시 좀 떨어집니다.      2

파일명  \
33                     [시승기] 기아 2021 K3.docx   
515                     기아 스포티지 리뷰(클리앙).docx   
373             Carnival_hybrid_Youtube.docx   
235  쌍용차 리스펙 티볼리, 인포콘 하나로 체질을 바꾼 소형 SUV.docx   
122             모든 것이 변했다, 쌍용 뉴 코란도 시승기.docx   

                                                    문장  label  
33   대신 가운데 디스플레이 테두리 안쪽 아래에 각종 기능 버튼이 터치식으로 자리를 잡으...      5  
515         특히 전 바늘계기판을 좋아하다 보니 소나타의 실내의 만족도 무척 높았습니다.      5  
373                  그리고 앞차와의 거리도 요즘 헤더 디스플레이 통해서 보여줘요      4  
235  한편 리스펙 티볼리는 또 스마트 폰 미러링, 10.25인치 풀 디지털 클러스터, 9...      4  
122                  가장 먼저 눈에 띄는 부분은 10.25인치의 디지털 계기판.      3

## Load the pre-trained model

### Subtask:
Load the KcBERT base model and tokenizer.


**Reasoning**:
Import the necessary classes and load the tokenizer and model for KcBERT.



In [ ]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification

model_name = "beomi/kcbert-base"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=6)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at beomi/kcbert-base and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


## Prepare training data for fine-tuning

### Subtask:
Tokenize and format the training data for input into the model.


**Reasoning**:
Define a function to tokenize the text and encode the labels, then apply it to the training and testing DataFrames and convert them to Hugging Face Dataset objects.



In [ ]:
from datasets import Dataset

def tokenize_function(row):
    inputs = tokenizer(row['문장'], padding='max_length', truncation=True)
    inputs['labels'] = row['label']
    return inputs

train_dataset = Dataset.from_pandas(train_df)
test_dataset = Dataset.from_pandas(test_df)

tokenized_train_dataset = train_dataset.map(tokenize_function, batched=True)
tokenized_test_dataset = test_dataset.map(tokenize_function, batched=True)

print("Tokenized training dataset:")
print(tokenized_train_dataset)
print("\nTokenized testing dataset:")
print(tokenized_test_dataset)

Map:   0%|          | 0/424 [00:00<?, ? examples/s]

Map:   0%|          | 0/106 [00:00<?, ? examples/s]

Tokenized training dataset:
Dataset({
    features: ['파일명', '문장', 'label', '__index_level_0__', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
    num_rows: 424
})

Tokenized testing dataset:
Dataset({
    features: ['파일명', '문장', 'label', '__index_level_0__', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
    num_rows: 106
})


## Fine-tune the model

### Subtask:
Train the KcBERT base model on the training set.


**Reasoning**:
Import the necessary classes and train the KcBERT model on the training set using the Hugging Face Trainer with the specified training arguments.



In [ ]:
from transformers import TrainingArguments, Trainer

training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=3,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    warmup_steps=500,
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
    eval_strategy="epoch",
    report_to="none",
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train_dataset,
    eval_dataset=tokenized_test_dataset,
)

trainer.train()

## Save the fine-tuned model

### Subtask:
Save the fine-tuned model to "/content/kcbert_selftuned".


**Reasoning**:
Save the fine-tuned model and tokenizer to the specified directory.



In [ ]:
output_dir = "/content/kcbert_selftuned"
trainer.save_model(output_dir)
tokenizer.save_pretrained(output_dir)

print(f"Fine-tuned model and tokenizer saved to {output_dir}")